FEA Group aware Time-based cross validation

getgaurav2 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Fixes #14257 
1 . Assumes that `groups` are contiguous . 
2. Split the `groups` into train and test indices  using TimeSeriesSplit 
3. Use these group indices to get the indices for original data.
4. Use the max_train_size parameter to trim the train_array  for each iteration of split.
#### Any other comments?

Question: 
1. Should we have separate parameter  for max_train_size in group vs. actual data ?
2. Should we have separate parameter  for gap in group vs. actual data ?



<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jnothman commented:
@getgaurav2, tests may be failing. Let us know when you want review.

jnothman commented:
Linting is failing, actually
.

getgaurav2 commented:
> @getgaurav2, tests may be failing. Let us know when you want review.

Sure @jnothman  . Will do . Thank you .

getgaurav2 commented:
@jnothman  Could you please opine on the approach that I have taken for this feature ?
I can work on performance metrics , documentation etc. if you are ok with the general direction .
Thanks

getgaurav2 commented:
>@jnothman @ogrisel  Could you please opine on the approach that I have taken for this feature ? ...

@getgaurav2 - gentle reminder  @ogrisel  

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

getgaurav2 commented:
> ...

@ogrisel @jnothman - can you pls review this PR ? Thank you 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2180,3 +2181,54 @@ def _build_repr(self):
     return '%s(%s)' % (class_name, _pprint(params, offset=len(class_name)))
+
+
+class GroupTimeSeriesSplit(TimeSeriesSplit):
```
Please add a docstring describing expected functionality.

getgaurav2 replied:
@jnothman - Added  the docstring . Can you please check .

getgaurav2 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1527,3 +1528,81 @@ def test_random_state_shuffle_false(Klass):
+
+def test_group_time_series_fail_groups_are_none():
+    """
+    The GroupTimeSeriesSplit with no group should work exactlhy as a
```
Yes , I will remove this in the next PR

albertvillanova commented:
@getgaurav2 are you still working on this PR?

getgaurav2 commented:
>@getgaurav2 are you still working on this PR?

@albertvillanova - yes . I would like to finish this please. Have been busy at work lately . Will find some time this week to make some progress . Thanks 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

albertvillanova commented:
@getgaurav2 Great! Because I do really need this feature in the next release ;)

Pimpwhippa commented:
hi @getgaurav2, I tried pushing here but I got an error remote: Permission to getgaurav2/scikit-learn.git denied to Pimpwhippa. My git remote -v are correct I think. Please could you check if you allow edits to this PR. Thank you.

albertvillanova commented:
@getgaurav2, as you have seen, it is not a good idea to rebase onto master during a Pull Request. If you would like to sync branches, it is better you make a merge:
```
git merge upstream master
```
Anyway, it does not seem necessary to do it: see GitHub message: *This branch has no conflicts with the base branch*.

getgaurav2 commented:
> hi @getgaurav2, I tried pushing here but I got an error remote: Permission to getgaurav2/scikit-learn.git denied to Pimpwhippa. My git remote -v are correct I think. Please could you check if you allow edits to this PR. Thank you.

<img width="934" alt="Screen Shot 2020-09-06 at 5 33 17 PM" src="https://user-images.githubusercontent.com/4666017/92335682-2fa71c00-f067-11ea-9a05-bded8dd6f070.png">

getgaurav2 commented:
>@getgaurav2 Great! Because I do really need this feature in the next release ;)

@albertvillanova - would you want to check the latest code base ? 
I want to talk through the best way to handle cases when the function may yield empty training splits  . 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Pimpwhippa commented:
@getgaurav2 thank you. I just opened a PR.

albertvillanova commented:
Fast check: why this empty file? `sklearn/model_selection/conftest.py`

albertvillanova commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1619,3 +1620,79 @@ def test_random_state_shuffle_false(Klass):
+    X = [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10], [11, 12], [13, 14]]
+
+    # Should fail if the 'groups' is None
+    assert_raises_regexp(ValueError,
```
This is using the `unittest` testing framework, which I think we no longer use. New tests should be implemented using the current new test framework: `pytest`. Maybe some other person can confirm this.

Just look at the test above this one, to see how to implement it.

getgaurav2 replied:
Done

albertvillanova commented:
I would like to point out an issue in your implementation. I have created a new test case:
```
groups = np.array(['a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'])
gtss = GroupTimeSeriesSplit(n_splits=3)
```
and this is what I get:
```ipython
In [49]: for train_idx, test_idx in gtss.split(groups, groups=groups):
    ...:     print(train_idx, test_idx)
    ...:     print(groups[train_idx], groups[test_idx])
    ...:     print()
    ...:
[0 1 2 3 4 5] [6 7 8 9]
['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b']

[0 1 2 3 4 5] [10 11 12 13]
['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'c' 'c' 'c']

[ 0  1  2  3  4  5  6  7  8  9 10] [14 15 16 17]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'd' 'd' 'd']
```
As you can see, there are **gaps** between the train and the test indices, i.e. they are not contiguous.

I would expect the following result:
```ipython
In [49]: for train_idx, test_idx in gtss.split(groups, groups=groups):
    ...:     print(train_idx, test_idx)
    ...:     print(groups[train_idx], groups[test_idx])
    ...:     print()
    ...:
[0 1 2 3 4 5] [6 7 8 9 10]
['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b' 'b']

[0 1 2 3 4 5 6 7 8 9 10] [11 12 13 14]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c']

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14] [15 16 17]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd' 'd']
```

By the way, and concerning the *gaps*, if you look at the next v0.24 release version of `TimeSeriesSplit`, there are two new parameters: `test_size` and `gap`. Maybe these should also be implemented in `GroupTimeSeriesSplit`.

Pimpwhippa commented:
'In each split, test indices must be higher than before, and thus shuffling  in cross validator is inappropriate'.
You made the class description clearer already but does that mean i can still shuffle the groups, but not the time right?

getgaurav2 commented:
> Fast check: why this empty file? `sklearn/model_selection/conftest.py`



> Fast check: why this empty file? `sklearn/model_selection/conftest.py`

removed

Pimpwhippa commented:
i mean this is allowed
groups = ['B', 'D', 'D', 'C', 'C', 'A', 'B', 'A']
and not only this
groups = ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D']

getgaurav2 commented:
> I would like to point out an issue in your implementation. I have created a new test case:
> 
> ```
> groups = np.array(['a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'])
> gtss = GroupTimeSeriesSplit(n_splits=3)
> ```
> 
> and this is what I get:
> 
> ```
> In [49]: for train_idx, test_idx in gtss.split(groups, groups=groups):
>     ...:     print(train_idx, test_idx)
>     ...:     print(groups[train_idx], groups[test_idx])
>     ...:     print()
>     ...:
> [0 1 2 3 4 5] [6 7 8 9]
> ['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b']
> 
> [0 1 2 3 4 5] [10 11 12 13]
> ['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'c' 'c' 'c']
> 
> [ 0  1  2  3  4  5  6  7  8  9 10] [14 15 16 17]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'd' 'd' 'd']
> ```
> 
> As you can see, there are **gaps** between the train and the test indices, i.e. they are not contiguous.
> 
> I would expect the following result:
> 
> ```
> In [49]: for train_idx, test_idx in gtss.split(groups, groups=groups):
>     ...:     print(train_idx, test_idx)
>     ...:     print(groups[train_idx], groups[test_idx])
>     ...:     print()
>     ...:
> [0 1 2 3 4 5] [6 7 8 9 10]
> ['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b' 'b']
> 
> [0 1 2 3 4 5 6 7 8 9 10] [11 12 13 14]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c']
> 
> [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14] [15 16 17]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd' 'd']
> ```
> 
> By the way, and concerning the _gaps_, if you look at the next v0.24 release version of `TimeSeriesSplit`, there are two new parameters: `test_size` and `gap`. Maybe these should be also implemented in `GroupTimeSeriesSplit`.

@albertvillanova - Thanks for your  review. I have incorporated your feedback  in the latest Commit . Can you please check again .

albertvillanova commented:
@getgaurav2, thanks for your contribution. Your implementation now gives the expected result in the case I pointed out above.

However, when using the new parameters (`gap`, `test_size`) I think I have found some unexpected behavior (that I am commenting below), besides that they are not documented in the docstring. Moreover, I think there are some peculiarities when dealing with groups that might deserve a discussion to decide the expected behavior, the definition of the corresponding API and its implementation. For example, I am thinking of `test_size`, which in `TimeSeriesSplit` corresponds to the number of *samples* in the test set; however, when dealing with groups, maybe it would be more useful to be able to specify the number of *groups* in the test set instead.

Maybe, I would suggest, forgetting about this parameters for the moment, so that this PR can already be finished. You could eventually open a new PR to implement those parameters. What do you think?

The unexpected behavior for the parameter `gap`. Current behavior:
```ipython
In [10]: gtss = GroupTimeSeriesSplit(n_splits=2, test_size=2, gap=1)

In [11]: for train_idx, test_idx in gtss.split(groups, groups=groups):
    ...:     print(train_idx, test_idx)
    ...:     print(groups[train_idx], groups[test_idx])
    ...:     print()
    ...:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c']

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14] [15, 16]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd']
```
As you can see, there is no gap between training and test sets.

The unexpected behavior for the parameter `test_size`. Current behavior:
```ipython
In [8]: gtss = GroupTimeSeriesSplit(n_splits=2, test_size=5)

In [9]: for train_idx, test_idx in gtss.split(groups, groups=groups):
   ...:     print(train_idx, test_idx)
   ...:     print(groups[train_idx], groups[test_idx])
   ...:     print()
   ...:
scikit-learn\sklearn\model_selection\_split.py:2375: UserWarning: The size=4 of group=c is smaller than test_size=5.
  warnings.warn(("The size=%d of group=%s is smaller"
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12, 13, 14]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c']

scikit-learn\sklearn\model_selection\_split.py:2375: UserWarning: The size=3 of group=d is smaller than test_size=5.
  warnings.warn(("The size=%d of group=%s is smaller"
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14] [15, 16, 17]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd' 'd']

```
In this case, I think the result should be:
```ipython
[0 1 2 3 4 5] [6 7 8 9 10]
['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b' 'b']

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12, 13, 14, 15]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c', 'd']
```

albertvillanova commented on sklearn/model_selection/_split.py:
```diff
@@ -2236,10 +2237,150 @@ def _build_repr(self):
 
 
+class GroupTimeSeriesSplit(_BaseKFold):
+    """Grouped Time Series cross-validator
```
```suggestion
    """Time Series cross-validator variant with non-overlapping groups.
```

albertvillanova commented on sklearn/model_selection/_split.py:
```diff
@@ -2236,10 +2237,150 @@ def _build_repr(self):
+    --------
+    >>> import numpy as np
+    >>> from sklearn.model_selection import GroupTimeSeriesSplit
+    >>> groups = ['a' ,'a' ,'a' ,'b','b' , 'b','c' ,'c' ,'c' ,'d' ,'d' ,'d']
```
```suggestion
    >>> groups = ['a', 'a', 'a', 'b', 'b', 'b', 'c', 'c', 'c', 'd', 'd', 'd']
```

albertvillanova commented on sklearn/model_selection/_split.py:
```diff
@@ -2236,10 +2237,150 @@ def _build_repr(self):
+    ...     print("TRAIN:", train_index, "TEST:", test_index)
+    TRAIN: [0, 1, 2, 3, 4, 5] TEST: [6, 7, 8]
+    TRAIN: [0, 1, 2, 3, 4, 5, 6, 7, 8] TEST: [9, 10, 11]
+    """
```
I would add here, besides the indices, also the groups (as we have done in the discussion), so that it is much clearer.

getgaurav2 replied:
Done .

albertvillanova commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1622,6 +1623,113 @@ def test_random_state_shuffle_false(Klass):
 
 
+def test_group_time_series_fail_groups_are_none():
+    """
+    The GroupTimeSeriesSplit with no group should raise an Error
+    """
```
I think we do not use docstrings for tests.
```suggestion
    # The GroupTimeSeriesSplit with no group should raise an Error
```

getgaurav2 replied:
Done

albertvillanova commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1622,6 +1623,113 @@ def test_random_state_shuffle_false(Klass):
+    train, test = next(splits)
+    assert_array_equal(train, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13])
+    assert_array_equal(test, [15, 16])
+
```
2 blank lines between tests.
```suggestion


```

getgaurav2 replied:
2 blank lines seem to give a linting error.

albertvillanova commented:
Maybe I would also add the test case I pointed out above:
```ipython
In [47]: groups = np.array(['a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'])
In [48]: gtss = GroupTimeSeriesSplit(n_splits=3)

In [49]: for train_idx, test_idx in gtss.split(groups, groups=groups):
    ...:     print(train_idx, test_idx)
    ...:     print(groups[train_idx], groups[test_idx])
    ...:     print()
    ...:
[0 1 2 3 4 5] [6 7 8 9 10]
['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b' 'b']

[0 1 2 3 4 5 6 7 8 9 10] [11 12 13 14]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c']

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14] [15 16 17]
['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd' 'd']
```

getgaurav2 commented:
> @getgaurav2, thanks for your contribution. Your implementation now gives the expected result in the case I pointed out above.
> 
> However, when using the new parameters (`gap`, `test_size`) I think I have found some unexpected behavior (that I am commenting below), besides that they are not documented in the docstring. Moreover, I think there are some peculiarities when dealing with groups that might deserve a discussion to decide the expected behavior, the definition of the corresponding API and its implementation. For example, I am thinking of `test_size`, which in `TimeSeriesSplit` corresponds to the number of _samples_ in the test set; however, when dealing with groups, maybe it would be more useful to be able to specify the number of _groups_ in the test set instead.
> 
> Maybe, I would suggest, forgetting about this parameters for the moment, so that this PR can already be finished. You could eventually open a new PR to implement those parameters. What do you think?
> 
> The unexpected behavior for the parameter `gap`. Current behavior:
> 
> ```
> In [10]: gtss = GroupTimeSeriesSplit(n_splits=2, test_size=2, gap=1)
> 
> In [11]: for train_idx, test_idx in gtss.split(groups, groups=groups):
>     ...:     print(train_idx, test_idx)
>     ...:     print(groups[train_idx], groups[test_idx])
>     ...:     print()
>     ...:
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c']
> 
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14] [15, 16]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd']
> ```
> 
> As you can see, there is no gap between training and test sets.
> 
> The unexpected behavior for the parameter `test_size`. Current behavior:
> 
> ```
> In [8]: gtss = GroupTimeSeriesSplit(n_splits=2, test_size=5)
> 
> In [9]: for train_idx, test_idx in gtss.split(groups, groups=groups):
>    ...:     print(train_idx, test_idx)
>    ...:     print(groups[train_idx], groups[test_idx])
>    ...:     print()
>    ...:
> scikit-learn\sklearn\model_selection\_split.py:2375: UserWarning: The size=4 of group=c is smaller than test_size=5.
>   warnings.warn(("The size=%d of group=%s is smaller"
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12, 13, 14]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c']
> 
> scikit-learn\sklearn\model_selection\_split.py:2375: UserWarning: The size=3 of group=d is smaller than test_size=5.
>   warnings.warn(("The size=%d of group=%s is smaller"
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14] [15, 16, 17]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] ['d' 'd' 'd']
> ```
> 
> In this case, I think the result should be:
> 
> ```
> [0 1 2 3 4 5] [6 7 8 9 10]
> ['a' 'a' 'a' 'a' 'a' 'a'] ['b' 'b' 'b' 'b' 'b']
> 
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12, 13, 14, 15]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c', 'd']
> ```

@albertvillanova  - Thanks for your feedback . 

- For the test_size = 5 , the expected output that you have put is certainly more desirable .  It will be interesting to determine  how it  should behave when there is  conflict between gap and test_size like  :  

groups = np.array(['a', 'a', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd', 'd', 'd'])
gtss = GroupTimeSeriesSplit(n_splits=2, test_size=5)

> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [11, 12, 13, 14, 15]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['c' 'c' 'c' 'c', 'd']
OR 
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] [15, 16, 17, 18, 19]
> ['a' 'a' 'a' 'a' 'a' 'a' 'b' 'b' 'b' 'b' 'b'] ['d' 'd' 'd' 'd', 'd']

- Would you recommend that I open another PR here or raise a new feature request to keep things separate ?

albertvillanova commented:
@getgaurav2, as you have already done an amazing amount of work here (thank you), my suggestion would be to leave for another PR the implementation of the parameters `test_size` and `gap`. That way, you would already be able to finish this PR.

Just a minor addition: could you please add `GroupTimeSeriesSplit` to the test `test_yields_constant_splits`?

And please, add a whatsnew entry to the file `v0.24.rst`, in the *model_selection* section, as a |MajorFeature|, and reference this PR and your name and link. You can take the other entries as example.

getgaurav2 commented:
> @getgaurav2, as you have already done an amazing amount of work here (thank you), my suggestion would be to leave for another PR the implementation of the parameters `test_size` and `gap`. That way, you would already be able to finish this PR.
> 
> Just a minor addition: could you please add `GroupTimeSeriesSplit` to the test `test_yields_constant_splits`?
> 
> And please, add a whatsnew entry to the file `v0.24.rst`, in the _model_selection_ section, as a |MajorFeature|, and reference this PR and your name and link. You can take the other entries as example.

@albertvillanova - Thank you . Can you please check now .

albertvillanova commented:
@getgaurav2 thanks. I think your PR is ready for review. Could you please change its name from WIP to MRG?

getgaurav2 commented:
> @getgaurav2 thanks. I think your PR is ready for a review round. Could you please change its name from WIP to MRG?

@albertvillanova  - would you  be able to add a label "waiting for reviewer" label here ?

albertvillanova commented:
> @albertvillanova - would you be able to add a "waiting for reviewer" label here ?

Maybe @cmarmo could do it?

getgaurav2 commented:
@ogrisel  , @jnothman  - Can you pls take look at this PR .  The logic seems to be popular on a recent Kaggle Competition thanks to the notebook from  @jorjinsmit ( https://www.kaggle.com/jorijnsmit/found-the-holy-grail-grouptimeseriessplit) which has since been referred by many others.
Thanks

jnothman commented on doc/whats_new/v0.24.rst:
```diff
@@ -406,13 +406,17 @@ Changelog
 :mod:`sklearn.model_selection`
 ..............................
 
+- |MajorFeature| :class:`model_selection.GroupTimeSeriesSplit` is a Time Series
```
this needs to move to the next release (doc/whats_new/v1.0.rst)

getgaurav2 replied:
Done

jnothman reviewed (commented):
Hi @getgaurav2, sorry for the delay. As far as I can tell, this PR doesn't test behaviour when groups are intermixed, e.g. `[1,1,1,1,2,2,1,1,3,3,3,3,4,3,3]`. If this isn't valid for your splitting strategy, it needs to be explicit in documentation.

Sorry to not give a more thorough review, but I'd like expected behaviour to be clarified through tests.

The user guide also needs updating in `doc/modules/cross_validation.rst`.

Pimpwhippa commented:
Hi @getgaurav2 

according to jnothman's comment,

> Sorry to not give a more thorough review, but I'd like expected behaviour to be clarified through tests.

regarding the tests, I'll try to write one more test that says like my question earlier

> i mean this is allowed
> groups = ['B', 'D', 'D', 'C', 'C', 'A', 'B', 'A']
> and not only this
> groups = ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D']

is that the idea?

getgaurav2 commented:
> Hi @getgaurav2
> 
> according to jnothman's comment,
> 
> > Sorry to not give a more thorough review, but I'd like expected behaviour to be clarified through tests.
> 
> regarding the tests, I'll try to write one more test that says like my question earlier
> 
> > i mean this is allowed
> > groups = ['B', 'D', 'D', 'C', 'C', 'A', 'B', 'A']
> > and not only this
> > groups = ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D']
> 
> is that the idea?

@Pimpwhippa - The idea would be not  to allow intermixed groups like `groups = ['B', 'D', 'D', 'C', 'C', 'A', 'B', 'A']`   in the input. I'll raise an error for this condition in the `GroupTimeSeriesSplit`  function and you can write a new test case to catch that error. We'll settle on the exact language of the error shortly . Please raise a PR on my branch like you did earlier. - Thanks

getgaurav2 commented:
@Pimpwhippa  - can you pls add a test case . The error should be : 
`ValueError: The groups should be continuous. Found a non-countinuous group at index=11`

Pimpwhippa commented:
@getgaurav2 sorry I didn't mean to commit directly to this PR. I wanted to open another PR to your branch as before, but when I pushed to your branch it comes directly here. Maybe because the settings have changed. Please let me know if I should revert it.

Pimpwhippa commented:
@getgaurav2 I'm terribly sorry I messed up your branch. Now it's time I'd better ask for help. Let me try to explain to you what I did. 

First I tried to go back to your commit 61dcb38 but I used git revert wrongly and it reverted also the 61dcb38 so I reverted the revert. Then I tried to remove the 3 commits I made here, one of which is a merge commit that is supposed to merge your branch onto my branch (115be74). So I reverted it but it turned out many things I didn't do are also reverted. Your latest changes in _split.py are also not there (why?)

So now I'm not sure what to do. Can I still try to go back to 61dcb38 with git revert --no-commit d041ff7..HEAD?

Your advice will be invaluable. Thank you very much.

getgaurav2 commented:
> @getgaurav2 I'm terribly sorry I messed up your branch. Now it's time I'd better ask for help. Let me try to explain to you what I did.
> 
> First I tried to go back to your commit [61dcb38](https://github.com/scikit-learn/scikit-learn/commit/61dcb389301af21c189a7cad4a156d3ec1b83510) but I used git revert wrongly and it reverted also the [61dcb38](https://github.com/scikit-learn/scikit-learn/commit/61dcb389301af21c189a7cad4a156d3ec1b83510) so I reverted the revert. Then I tried to remove the 3 commits I made here, one of which is a merge commit that is supposed to merge your branch onto my branch ([115be74](https://github.com/scikit-learn/scikit-learn/commit/115be740286d440fad7017b0deeb5d6f139d077b)). So I reverted it but it turned out many things I didn't do are also reverted. Your latest changes in _split.py are also not there (why?)
> 
> So now I'm not sure what to do. Can I still try to go back to [61dcb38](https://github.com/scikit-learn/scikit-learn/commit/61dcb389301af21c189a7cad4a156d3ec1b83510) with git revert --no-commit [d041ff7](https://github.com/scikit-learn/scikit-learn/commit/d041ff79be97faf4c41b92a49d51a9d4dd7baa29)..HEAD?
> 
> Your advice will be invaluable. Thank you very much.

No Problem . Let me check . Also , I am looking at  the test case that you are trying to add  . I am assuming you will add an assert statement at some point.

Pimpwhippa commented:
@getgaurav2  Thank you very much for your help. Now i think i understand what i did wrong. Everything i thought i did to open a pull request to your branch was actually on your branch on your pull request to scikit-learn. 

Is it an idea also to checkout 61dcb38 and create a new branch and just push the new branch to yours? (but im not doing anything right now)

Im not sure about the assert statement. I thought with pytest raise(match='ValueError') will already does the job. I can check that out.

getgaurav2 commented:
> @getgaurav2 Thank you very much for your help. Now i think i understand what i did wrong. Everything i thought i did to open a pull request to your branch was actually on your branch on your pull request to scikit-learn.
> 
> Is it an idea also to checkout [61dcb38](https://github.com/scikit-learn/scikit-learn/commit/61dcb389301af21c189a7cad4a156d3ec1b83510) and create a new branch and just push the new branch to yours? (but im not doing anything right now)
> 
> Im not sure about the assert statement. I thought with pytest raise(match='ValueError') will already does the job. I can check that out.

@Pimpwhippa  - can you please add the test cases in the comments here ?

Pimpwhippa commented:
def test_group_time_series_is_contiguous():
    #test that groups are contiguous and not intermixed
    n_samples = _num_samples(X)
    groups = np.array(['b', 'd', 'd', 'c', 'c', 'a', 'b', 'a', 'b', 'd'])
    #groups = np.array(['d', 'd', 'd', 'c', 'c', 'c', 'a', 'a', 'a', 'a'])
    group_dict = {}
    for idx in np.arange(n_samples):
        if (groups[idx] in group_dict):
            if (idx - group_dict[groups[idx]][-1] == 1):
                group_dict[groups[idx]].append(idx)
            else:
                with pytest.raises(ValueError, match="The groups should be contiguous."
                    " Found a non-contiguous group at"
                    " index={0}".format(idx)):
                    raise ValueError("The groups should be contiguous."
                    " Found a non-contiguous group at"
                    " index={0}".format(idx))    
        else:
            group_dict[groups[idx]] = [idx]`

@getgaurav2 I think the word is contiguous? it means adjacent, bordering.

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2241,6 +2241,159 @@ def _build_repr(self):
+    Note that unlike standard cross-validation methods, successive
+    training sets are supersets of those that come before them.
+
+    The groups should be continuous. For Example:
```
# Documenting that groups should be continuous.

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2241,6 +2241,159 @@ def _build_repr(self):
+
+    Note that unlike standard cross-validation methods, successive
+    training sets are supersets of those that come before them.
+
```
Documenting that the group should be continuous.

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2241,6 +2241,159 @@ def _build_repr(self):
+        n_groups = _num_samples(unique_groups)
+        for idx in np.arange(n_samples):
+            if (groups[idx] in group_dict):
+                if (idx - group_dict[groups[idx]][-1] == 1):
```
Raise an error if the groups are non-continous.

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2241,6 +2241,159 @@ def _build_repr(self):
+        n_samples = _num_samples(X)
+        n_groups = _num_samples(unique_groups)
+        for idx in np.arange(n_samples):
+            if (groups[idx] in group_dict):
```
Raise an error if the groups are non-continuous.

getgaurav2 commented on doc/whats_new/v1.0.rst:
```diff
@@ -47,6 +47,10 @@ Changelog
 :mod:`sklearn.cluster`
 ......................
 
```
Moved to the correct file.

getgaurav2 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1728,6 +1728,18 @@ def test_group_time_series_non_overlap_group():
     assert_array_equal(groups[test], ['d', 'd', 'd'])
 
 
```
Add test case to check for error for non-continuous groups.

getgaurav2 commented:
> ```
> def test_group_time_series_is_contiguous():
> #test that groups are contiguous and not intermixed
> n_samples = _num_samples(X)
> groups = np.array(['b', 'd', 'd', 'c', 'c', 'a', 'b', 'a', 'b', 'd'])
> #groups = np.array(['d', 'd', 'd', 'c', 'c', 'c', 'a', 'a', 'a', 'a'])
> group_dict = {}
> for idx in np.arange(n_samples):
>     if (groups[idx] in group_dict):
>         if (idx - group_dict[groups[idx]][-1] == 1):
>             group_dict[groups[idx]].append(idx)
>         else:
>             with pytest.raises(ValueError, match="The groups should be contiguous."
>                 " Found a non-contiguous group at"
>                 " index={0}".format(idx)):
>                 raise ValueError("The groups should be contiguous."
>                 " Found a non-contiguous group at"
>                 " index={0}".format(idx))    
>     else:
>         group_dict[groups[idx]] = [idx]`
> ```
> 
> @getgaurav2 May I change the word to contiguous?

@Pimpwhippa  - Can you please check the latest commit . Calling the function within the test case would allow for better code maintenance and  ensure code coverage does not decline.

Pimpwhippa commented:
@getgaurav2 oh, it's that simple! I tried calling the function within the test case but somehow with next() the index keeps moving. Thank you so much for your advice.

getgaurav2 commented:
> Hi @getgaurav2, sorry for the delay. As far as I can tell, this PR doesn't test behaviour when groups are intermixed, e.g. `[1,1,1,1,2,2,1,1,3,3,3,3,4,3,3]`. If this isn't valid for your splitting strategy, it needs to be explicit in documentation.
> 
> Sorry to not give a more thorough review, but I'd like expected behaviour to be clarified through tests.
> 
> The user guide also needs updating in `doc/modules/cross_validation.rst`.

@jnothman  - I have made the changes in the code and documentation . Working on `doc/modules/cross_validation.rst` .

I had to reset the code to a previous commit and add the entire function again  . The real change since  your  prev.  review is the section that raises error if the groups are non-continuous.

getgaurav2 commented:
Looks like the Build is failing at present due to #18930 . 

@jnothman  I have made the changes  . The picture in cross-validation.html for `GroupTimeSeriesSplit` looks same as `TimeSeriesSplit` . I have used the same convention as was used in  plot_cv_indices.py for other cross validators.  Please let me know if you  think  it will be more useful to use some other data points to emphasize difference between these  two classes  . For example: 
<img width="421" alt="Screen Shot 2021-01-03 at 9 43 25 PM" src="https://user-images.githubusercontent.com/4666017/103496598-da102f80-4e0c-11eb-9af0-bd274a55f6de.png">

jnothman commented:
> Please let me know if you think it will be more useful to use some other data points to emphasize difference between these two classes . For example:

Yes, your groups need non-equal sizes.

getgaurav2 commented:
> > Please let me know if you think it will be more useful to use some other data points to emphasize difference between these two classes . For example:
> 
> Yes, your groups need non-equal sizes.

Like this ?
<img width="472" alt="Screen Shot 2021-01-15 at 9 58 23 AM" src="https://user-images.githubusercontent.com/4666017/104742314-4564d600-5718-11eb-9f66-4fbd70518d69.png">

tczhao commented:
Sorry for not knowing this PR in the first place.

Perhaps allowing supply group parameter in original TimeSeriesSplit.split is a cleaner approach, compare to having a branch new class. #19496

What do you think?

Pimpwhippa commented:
@tczhao I just remember to come check your comment now. yes, I can agree that your approach to use group parameter in the original TimeSeriesSplit.split indeed sounds cleaner. but this is not really my position to say. this PR is owned by @getgaurav2. not sure if he has seen your comment and your PR. usually he reacts quickly.

glemaitre commented:
> I can agree that your approach to use group parameter in the original TimeSeriesSplit.split indeed sounds cleaner. but this is not really my position to say.

I don't think this is appropriate because it would not follow the pattern used for the other CV

jorijnsmit commented:
> > I can agree that your approach to use group parameter in the original TimeSeriesSplit.split indeed sounds cleaner. but this is not really my position to say.
> 
> I don't think this is appropriate because it would not follow the pattern used for the other CV

Can you elaborate? I too believe that integrating `groups` into the original `TimeSeriesSplit` might be cleaner. What do you mean by "the pattern for the other CV"?

glemaitre commented:
In our API, we have the following CVs: `GroupKFold`, `GroupShuffleSplit`, `LeaveOneGroupOut`, `LeavePGroupsOut`. If we want to be consistent with those, we will need to go toward creating a new class.

We might want to consider a refactoring if all CVs classes support grouping samples and if there is a clear advantage in both maintenance and user experience of deprecating the `Group` classes in favor of introducing a `group` parameter only.

gialdetti commented:
Hi, 

To be consistent with the current naming convention (e.g., `LeaveOneGroupOut`), shouldn't it be named `TimeSeriesGroupSplit` - splitting ordered (by time) groups. Whereas the name `GroupTimeSeriesSplit` might imply splitting into groups of full-length time-series vectors.

I hope it makes sense.

Thanks,
Eyal

getgaurav2 commented:
From  @jnothman  's [comment](https://github.com/scikit-learn/scikit-learn/pull/19996#pullrequestreview-646654488) 
The assumption of continuous samples can effectively reduce the problem to running `TimeSeriesSplit` on the groups  then mapping the indices back to the original samples to create train and test splits .

fingoldo commented:
Hi all. getgaurav2 thank you for your efforts. I can hardly believe such an important and highly requested feature needs almost 2 years (and counting) to get released. When can sklearn users hope to expect it, any estimate, please?

glemaitre commented:
>  I can hardly believe such an important and highly requested feature needs almost 2 years (and counting) to get released. When can sklearn users hope to expect it, any estimate, please?

I can hardly believe that you did not propose yourself to contribute and finish up this effort  :)

fingoldo commented:
> > I can hardly believe such an important and highly requested feature needs almost 2 years (and counting) to get released. When can sklearn users hope to expect it, any estimate, please?
> 
> I can hardly believe that you did not propose yourself to contribute and finish up this effort :)

Thank you so much! I am so inexperienced with GitHub and all the teamwork. I only contributed to opensource once, to a much smaller project, and it was a nightmare!! Thank you so much Guillaume for working on this.

wmoreiraa commented:
Hi, @glemaitre @jnothman ! Is there anything I could do to help get this functionality rolled out?

ogrisel commented:
> In our API, we have the following CVs: GroupKFold, GroupShuffleSplit, LeaveOneGroupOut, LeavePGroupsOut. If we want to be consistent with those, we will need to go toward creating a new class.

> We might want to consider a refactoring if all CVs classes support grouping samples and if there is a clear advantage in both maintenance and user experience of deprecating the Group classes in favor of introducing a group parameter only.

Honestly, I think `KFold` and `ShuffleSplit` could accept a `group` parameter with default set to `None` and we could deprecate `GroupKFold` and `GroupShuffleSplit` to simplify the library.

For `LeaveOneGroupOut` and `LeavePGroupsOut` passing groups is mandatory, it makes no sense to use those without group info.

So I would not be opposed to adding an optional `group` parameter to `TimeSeriesSplit`.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2495,6 +2497,162 @@ def _build_repr(self):
+
+    test_size : int, default=None
+        Used to limit the size of the test set. Defaults to
+        ``n_samples // (n_splits + 1)``, which is the maximum allowed value
```
In a similar vein, this doesn't seem to be specified in the correct units?

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2495,6 +2497,162 @@ def _build_repr(self):
+        Number of splits. Must be at least 2.
+
+    max_train_size : int, default=None
+        Maximum size for a single training set.
```
Please specify the units: this size is the number of groups, isn't it?

getgaurav2 replied:
done

jnothman reviewed (commented):
A few nitpicks

glemaitre commented:
It seems that you got into trouble when merging `main` into your branch since there are a lot of changes.

@getgaurav2 Could you fix it?

ogrisel commented:
> It seems that you got into trouble when merging main into your branch since there are a lot of changes.

> @getgaurav2 Could you fix it?

It's mostly a py36 env folder that seems to have been added by mistake, isn't it?

Can you please delete it? It would help to have a clean diff to finalize the review.

Note that at the dev meeting we decided that it's ok to introduce a new class for the sake of consistency and that later we might consider refactoring all the splitters at once to consistently add the optional groups parameter to the original splitter (`KFolder`, `ShuffleSplit`, `TimeSeriesSplit`, ...) when it's not mandatory for the sake of simplification.

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2495,6 +2497,162 @@ def _build_repr(self):
+
+        for train_idx, test_idx in tss.split(unique_groups):
+            train_array = list(np.where(np.isin(groups, unique_groups[train_idx]))[0])
+            test_array = list(np.where(np.isin(groups, unique_groups[test_idx]))[0])
```
removed list.

getgaurav2 commented:
@ogrisel , @glemaitre  - Getting this `bash: build_tools/github/build_doc.sh: No such file or directory` error ...Looks like this is CircleCI related.

ogrisel commented:
> @ogrisel , @glemaitre - Getting this bash: build_tools/github/build_doc.sh: No such file or directory error ...Looks like this is CircleCI related.

I just pushed a merge of the `main` branch which was recently updated to fix circle ci related problems. Let's hope that this will also fix the problem on this PR.

getgaurav2 commented:
> > @ogrisel , @glemaitre - Getting this bash: build_tools/github/build_doc.sh: No such file or directory error ...Looks like this is CircleCI related.
> 
> I just pushed a merge of the `main` branch which was recently updated to fix circle ci related problems. Let's hope that this will also fix the problem on this PR.

It worked . Thanks @ogrisel .

lesteve commented:
There was no rendered documentation for your PR, I updated your branch, this should fix it and make it easier to review the documentation part of your PR.

ogrisel commented:
Thanks Loic for fixing the rendering. I also pushed another commit to improve the formatting of the docstring (using the [Rewrap](https://stkb.github.io/Rewrap/) VS Code extension).

I will do a proper review now.

ogrisel commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1883,15 +1884,171 @@ def test_random_state_shuffle_false(Klass):
+    with pytest.raises(
+        ValueError,
+        match=(
+            "The groups should be continuous. Found a non-countinuous group at index=15"
```
```suggestion
            "The groups should be continuous. Found a non-contiguous group at index=6"
```

ogrisel commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1883,15 +1884,171 @@ def test_random_state_shuffle_false(Klass):
+        ),
+    )
+    assert_array_equal(groups[test], ["d", "d", "d"])
+
```
Could you please add at least one assertion that checks a case where there is more than one group present in the test set? I think that should involve a larger number of unique groups with a small number of samples per group (e.g. groups of 10 groups of 2 or 3 data points each for instance).

getgaurav2 replied:
Done . Added new test cases test_group_time_series_non_overlap_group_**(1 , 2, 3 )**

ogrisel reviewed (commented):
Here is a pass of feedback.

ogrisel commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+    Note that unlike standard cross-validation methods, successive training
+    sets are supersets of those that come before them.
+
+    The group labels should be continuous such as the following:
```
```suggestion
    The group labels should be contiguous such as the following:
```

ogrisel commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+        'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'
+    ])
+
+    Non-continuous groups like below will give an error.
```
```suggestion
    Non-contiguous groups like below will give an error.
```

ogrisel commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1883,15 +1884,171 @@ def test_random_state_shuffle_false(Klass):
+    assert_array_equal(groups[test], ["d", "d", "d"])
+
+
+def test_group_time_series_non_continuous():
```
```suggestion
def test_group_time_series_non_contiguous():
```

ogrisel reviewed (commented):
More continuous => contiguous fixes.

ogrisel commented:
Also the diagram for the 2 time-aware splitters have time-order classes. This is very likely to never happen in a practical setting and would prevent any machine learning model to behave correctly on such a task and is therefore confusing to included in the documentation.

Therefore I would advise to randomly scatter the samples of the 3 classes across the samples, at least for those 2 splitters.

glemaitre commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+        unique_groups = u[np.argsort(ind)]
+        n_samples = _num_samples(X)
+        n_groups = _num_samples(unique_groups)
+        for idx in np.arange(n_samples):
+            if groups[idx] in group_dict:
+                if idx - group_dict[groups[idx]][-1] == 1:
+                    group_dict[groups[idx]].append(idx)
+                else:
+                    raise ValueError(
+                        (
+                            "The groups should be continuous."
+                            " Found a non-countinuous group at"
+                            " index={0}"
+                        ).format(idx)
+                    )
+            else:
+                group_dict[groups[idx]] = [idx]
+            if n_folds > n_groups:
+                raise ValueError(
+                    (
+                        "Cannot have number of folds={0} greater than"
+                        " the number of groups={1}"
+                    ).format(n_folds, n_groups)
```
I think that we can speed up this dictionary generation with NumPy:

```python
reordered_unique_groups, indices = np.unique(groups, return_index=True)
original_order_group = indices.argsort()
groups_indices = np.split(np.arange(groups.size), indices[original_order_group][1:])
group_dict = {
   g: i for g, i in zip(reordered_unique_groups[original_order_group], groups_indices)
}
```

We can then make a check to be sure that the indices are consistent.

glemaitre commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+            test_size=None,
+        )
+
+        for train_idx, test_idx in tss.split(unique_groups):
```
I find the behaviour of this split quite surprising.

Here, we discard the fact that the group can be imbalanced. Therefore, we could end up with a split where the different test sets could have very different sizes.

I would have intuitively expected the following:

- apply `tss` on the original data
- find the closest split group-aligned using `group_dict`

In this case, we ensure that the test splits will have approximately the same size. However, we might have a different number of groups in the different test sets.

@getgaurav2 Did we consider this approach and if we discarded it do you recall what is the advantage of one approach compared to the other one?

getgaurav2 replied:
@glemaitre  - I agree with your observation  . Originally , I had thought about approaching the problem the same way as you described. 

I had realized that , after step - 1 when elements of one group will split into both Train and Test then we will need some  function to  find the "closest" split which is group aligned .  Here is one example to start  the conversation : 
- Let's say after step - 1 the split looks like : 
 TRAIN : [ 'a' , 'a' , 'a', 'a', 'a' , 'b']   TEST : ['b', 'b' , 'b' , 'b' , 'c' , 'c']
- Then after step -2  should the split be : 
 TRAIN : [ 'a' , 'a' , 'a', 'a', 'a' ]   TEST : ['b', 'b', 'b' , 'b' , 'b' , 'c' , 'c']
Reason (intuitive) : lengths of TRAIN and TEST after step -1  are equal , therefore the 'b' from TRAIN shall move into  TEST .

In my mind , This function shall be a bunch of rules that consider  -  length of splits , gap , group sizes , max_train_length , max_test_length etc. 
I could not think of an  elegant way to handle the  overlapping groups while minimizing  the difference in the length of splits.

jnothman replied:
I don't think stratification is an appropriate objective given the other constraints here.

@glemaitre it's unclear to me what the algorithm is for finding "closest split group-aligned" that ensures each test set is distinct.

glemaitre reviewed (commented):
I think that we could an additional test with a bit more groups and imbalanced size such that we can make sure of the behaviour that I describe in one of the comments (equal test size/unequal group size vs. unequel test size/equal group size)

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+                    (
+                        "Cannot have number of folds={0} greater than"
+                        " the number of groups={1}"
+                    ).format(n_folds, n_groups)
```
done

getgaurav2 replied:
> an f-string as well here.

done

getgaurav2 commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+                            "The groups should be continuous."
+                            " Found a non-countinuous group at"
+                            " index={0}"
+                        ).format(idx)
```
> Could you use an f-string

done

getgaurav2 commented on examples/model_selection/plot_cv_indices.py:
```diff
@@ -194,12 +201,15 @@ def plot_cv_indices(cv, X, y, group, ax, n_splits, lw=10):
     this_cv = cv(n_splits=n_splits)
     fig, ax = plt.subplots(figsize=(6, 3))
+    if cv == GroupTimeSeriesSplit:
+        groups = unevengroups
```
I am now calling plot_cv_indices with different arguments ; `if cv == GroupTimeSeriesSplit then unevengroups else groups `.

getgaurav2 replied:
> Once passing the `GroupTimeSeriesSplit`, `groups` will always be set to `unevengroups`. We should have a local `groups variable then.

I am now calling plot_cv_indices with different arguments ; if cv == GroupTimeSeriesSplit then unevengroups else groups

getgaurav2 replied:
I am now calling plot_cv_indices with different arguments ; if cv == GroupTimeSeriesSplit then unevengroups else groups

getgaurav2 commented on sklearn/model_selection/__init__.py:
```diff
@@ -71,6 +70,7 @@
     "permutation_test_score",
     "train_test_split",
     "validation_curve",
+    "GroupTimeSeriesSplit",
```
done

cmarmo commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,150 @@ def _build_repr(self):
+    Note that, unlike standard cross-validation methods, successive training
+    sets are supersets of those that come before them.
+
+    The group labels should be contiguous such as the following:
+
+    valid_groups = np.array([
+        'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'
+    ])
```
```suggestion
    The group labels should be contiguous such as the following:

    .. code-block::

        valid_groups = np.array([
            'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'
        ])
```

cmarmo commented on sklearn/model_selection/_split.py:
```diff
@@ -2506,6 +2508,150 @@ def _build_repr(self):
+        'a', 'a', 'b', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c', 'd', 'd', 'd'
+    ])
+
+    Non-contiguous groups like below will give an error.
+
+    invalid_groups = np.array([
+        'a', 'a', 'b', 'b', 'b', 'b', 'b', 'a', 'c', 'c', 'c', 'b', 'd', 'd'
+    ])
```
```suggestion
    Non-contiguous groups like below will give an error.

    .. code-block::

        invalid_groups = np.array([
            'a', 'a', 'b', 'b', 'b', 'b', 'b', 'a', 'c', 'c', 'c', 'b', 'd', 'd'
        ])
```

cmarmo reviewed (commented):
Thanks @getgaurav2 for your patience.
Those two comments should fix the documentation error and rendering.

getgaurav2 commented:
> Also the diagram for the 2 time-aware splitters have time-order classes. This is very likely to never happen in a practical setting and would prevent any machine learning model to behave correctly on such a task and is therefore confusing to included in the documentation.
> 
> Therefore I would advise to randomly scatter the samples of the 3 classes across the samples, at least for those 2 splitters.

@ogrisel - 
1.  I cant' say I completely understand the difference between class and group. I hope I understood your comment correctly. Do the attached plots match what you have in mind.
2. I had mentioned this in an earlier comment but the plot for both `GroupTimeSeriesSplit` and  `TimeSeriesSplit` look exactly the same . Does it make sense to keep group in the regular `TimeSeriesSplit` ?
<img width="1200" alt="Screen Shot 2022-07-16 at 5 37 28 PM" src="https://user-images.githubusercontent.com/4666017/179373146-e92f1618-50ab-4905-8533-d25a87cffc9e.png">

getgaurav2 commented:
> I think that we could an additional test with a bit more groups and imbalanced size such that we can make sure of the behaviour that I describe in one of the comments (equal test size/unequal group size vs. unequel test size/equal group size)

Added a test case with imbalanced groups . Imbalanced splits still pass the test . I can change it if we agree on a function to that will give "closest group-aligned split" ( [comment )](https://github.com/scikit-learn/scikit-learn/pull/16236#discussion_r933811595)

he7d3r commented on doc/modules/cross_validation.rst:
```diff
@@ -902,6 +902,60 @@ Here is a visualization of the cross-validation behavior.
+we may not want the training days to be be used in testing. Here the days can
+act as groups to keep the training and test splits separate.
+
+Example of 3-split time series cross-validation on a dataset with 18 samples
```
```suggestion
Example of 3-split time series cross-validation on a dataset with 14 samples
```

he7d3r commented on sklearn/model_selection/_split.py:
```diff
@@ -2574,6 +2576,154 @@ def _build_repr(self):
+            raise ValueError("The 'groups' parameter should not be None.")
+        X, y, groups = indexable(X, y, groups)
+        n_samples, n_folds, group_dict = _num_samples(X), self.n_splits + 1, {}
+        # `np.unique` will reordered the group. We need to keep the original
```
```suggestion
        # `np.unique` will reorder the group. We need to keep the original
```

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2685,6 +2687,154 @@ def _build_repr(self):
+    In each split, test indices must be higher than before, and thus shuffling
+    in cross validator is inappropriate.
+
+    This cross-validation object is a variation of :class:`KFold`. In the kth
```
I don't think it helps the reader much to call this a variation of KFold. The only apparent similarity is that test sets are non-overlapping?

getgaurav2 replied:
Removed.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2685,6 +2687,154 @@ def _build_repr(self):
+                    )
+            else:
+                group_dict[groups[idx]] = [idx]
+            if n_folds > n_groups:
```
shouldn't this be out of the loop?

getgaurav2 replied:
Done . Thanks for catching

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2685,6 +2687,154 @@ def _build_repr(self):
+        if groups is None:
+            raise ValueError("The 'groups' parameter should not be None.")
+        X, y, groups = indexable(X, y, groups)
+        n_samples, n_folds, group_dict = _num_samples(X), self.n_splits + 1, {}
```
please avoid three separate initialisations on one line

getgaurav2 replied:
Done

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2685,6 +2687,154 @@ def _build_repr(self):
+        unique_groups = reordered_unique_groups[np.argsort(indices)]
+        n_samples = _num_samples(X)
+        n_groups = len(unique_groups)
+        for idx in np.arange(n_samples):
```
please use `range` rather than arange here since we do not use the array as an array.

getgaurav2 replied:
Done

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -2685,6 +2687,154 @@ def _build_repr(self):
+        for idx in np.arange(n_samples):
+            if groups[idx] in group_dict:
+                if idx - group_dict[groups[idx]][-1] == 1:
+                    group_dict[groups[idx]].append(idx)
```
we should not need to store full lists... I think this is sufficient:
```py
seen_groups = set()
prev_group = None
for group in groups:
    if group != prev_group and group in seen_groups:
        raise ValueError("The groups should be contiguous")
    prev_group = group
```

jnothman replied:
Alternatively, we could do:
```py
reordered_unique_groups, indices, inverse = np.unique(groups, return_index=True, return_inverse=True)
if (np.diff(inverse) < 1).any():
    raise ValueError("The groups should be contiguous")
```
although this is much more obfuscated!

getgaurav2 replied:
Used the first variant with `enumerate` to add the index in the error msg.

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -1883,15 +1884,1048 @@ def test_random_state_shuffle_false(Klass):
+    assert_array_equal(groups[test], ["d", "d", "d"])
+
+
+def test_group_time_series_non_overlap_group_2():
```
What is the benefit of having each of these test cases? If they test distinct capabilities, consider using `pytest.parametrize` and expressing your test in a more generalised way that tests the invariances of interest.

getgaurav2 replied:
I have renamed the test cases to convey the distinction. Also , removed one of the tests that was kind of repetitive.

jnothman reviewed (changes requested):
The tests are currently quite hard to read. Tests need to be easy to read and understand that they are testing what we believe to be true.

adrinjalali commented:
@glemaitre this seems like a good candidate for a "time series related" "project board".

lu0x1a0 commented:
Would I be correct that this PR would also close issue https://github.com/scikit-learn/scikit-learn/issues/6322 and related old PR https://github.com/scikit-learn/scikit-learn/pull/6351 's `HeteroTSCV`.? 

Also I feel like a gap functionality that is similar to timegapsplit https://github.com/scikit-learn/scikit-learn/pull/13761 and the current `gap` parameter in https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html would be useful when it comes to chronological groups.

adrinjalali commented:
@lu0x1a0 this is up for grabs. Now that we have metadata routing, this can nicely move forward if you're up for it.

getgaurav2 commented:
@adrinjalali - I would love to continue to see this one to completion if that's okay .

adrinjalali commented:
@getgaurav2 that'd be fantastic!

getgaurav2 commented:
> @getgaurav2 that'd be fantastic!

@adrinjalali - would you please be able to check the issue in doc build to and ive some direction ....thank you !

adrinjalali commented:
Something has gone wrong here probably in your merge with `main`. You have over 12k line change, and some large changes in the `rst` files. I'd start with fixing those issues if I were you.

getgaurav2 committed:
Clean Commit

getgaurav2 committed:
fix linting error with unused import

getgaurav2 committed:
Add change log

getgaurav2 committed:
plot grouptimeseries split

getgaurav2 committed:
linting error

getgaurav2 commented:
> Something has gone wrong here probably in your merge with `main`. You have over 12k line change, and some large changes in the `rst` files. I'd start with fixing those issues if I were you.

@adrinjalali  - sorry for the long gap . Would you pls be able to review now ?

getgaurav2 committed:
using reange and repeat to condense code

getgaurav2 committed:
ruff formatting

lcrmorin commented:
This is something of interest for panel data. Is there something that can be done to help ? @getgaurav2 @jnothman

getgaurav2 commented:
> This is something of interest for panel data. Is there something that can be done to help ? @getgaurav2 @jnothman 

I would appreciate a review and feedback on the code from the core team
